Mathlib/Tactic/SimpTrace.lean

```diff
@@ -0,0 +1,204 @@
+
+@[inheritDoc simpTrace]
+macro tk:"simp?!" rest:simpTraceArgsRest : tactic => `(tactic| simp?%$tk ! $rest)
+
+macro_rules
+  | `(tactic| simp?%$tk $(config)? $(discharger)? $[only%$o]? $[[$args,*]]? $(loc)?) =>
+    `(tactic| set_option tactic.simp.trace true in
+      simp%$tk $(config)? $(discharger)? $[only%$o]? $[[$args,*]]? $(loc)?)
+  | `(tactic| simp?%$tk ! $(config)? $(discharger)? $[only%$o]? $[[$args,*]]? $(loc)?) =>
+    `(tactic| set_option tactic.simp.trace true in
+      simp!%$tk $(config)? $(discharger)? $[only%$o]? $[[$args,*]]? $(loc)?)
+
+/-- The common arguments of `simp_all?` and `simp_all?!`. -/
+syntax simpAllTraceArgsRest := (config)? (discharger)? (&" only")? (dsimpArgs)?
+
+@[inheritDoc simpTrace]
+syntax (name := simpAllTrace) "simp_all?" "!"? simpAllTraceArgsRest : tactic
+
+@[inheritDoc simpTrace]
+macro tk:"simp_all?!" rest:simpAllTraceArgsRest : tactic => `(tactic| simp_all?%$tk ! $rest)
+
+macro_rules
+  | `(tactic| simp_all?%$tk $(config)? $(discharger)? $[only%$o]? $[[$args,*]]?) =>
+    `(tactic| set_option tactic.simp.trace true in
+      simp_all%$tk $(config)? $(discharger)? $[only%$o]? $[[$args,*]]?)
+  | `(tactic| simp_all?%$tk ! $(config)? $(discharger)? $[only%$o]? $[[$args,*]]?) =>
+    `(tactic| set_option tactic.simp.trace true in
+      simp_all!%$tk $(config)? $(discharger)? $[only%$o]? $[[$args,*]]?)
+
+/-- The common arguments of `dsimp?` and `dsimp?!`. -/
+syntax dsimpTraceArgsRest := (config)? (&" only")? (dsimpArgs)? (ppSpace location)?
+
+@[inheritDoc simpTrace]
+syntax (name := dsimpTrace) "dsimp?" "!"? dsimpTraceArgsRest : tactic
+
+@[inheritDoc simpTrace]
+macro tk:"dsimp?!" rest:dsimpTraceArgsRest : tactic => `(tactic| dsimp?%$tk ! $rest)
+
+macro_rules
+  | `(tactic| dsimp?%$tk $(config)? $[only%$o]? $[[$args,*]]? $(loc)?) =>
+    `(tactic| set_option tactic.simp.trace true in
+      dsimp%$tk $(config)? $[only%$o]? $[[$args,*]]? $(loc)?)
+  | `(tactic| dsimp?%$tk ! $(config)? $[only%$o]? $[[$args,*]]? $(loc)?) =>
+    `(tactic| set_option tactic.simp.trace true in
+      dsimp!%$tk $(config)? $[only%$o]? $[[$args,*]]? $(loc)?)
+
+/--
+`squeeze_scope a => tacs` is part of the implementation of `squeeze_scope`.
+Inside `tacs`, invocations of `simp` wrapped with `squeeze_wrap a _ => ...` will contribute
+to the accounting associated to scope `a`.
+-/
+local syntax (name := squeezeScopeIn) "squeeze_scope " ident " => " tacticSeq : tactic
+/--
+`squeeze_wrap a x => tac` is part of the implementation of `squeeze_scope`.
+Here `tac` will be a `simp` or `dsimp` syntax, and `squeeze_wrap` will run the tactic
+and contribute the generated `usedSimps` to the `squeezeScopes[a][x]` variable.
+-/
+local syntax (name := squeezeWrap) "squeeze_wrap " ident ident " => " tactic : tactic
+
+open TSyntax.Compat in
+/--
+The `squeeze_scope` tactic allows aggregating multiple calls to `simp` coming from the same syntax
+but in different branches of execution, such as in `cases x <;> simp`.
```

Comment:
Perhaps best to write `simp?` in both appearances here, just in case someone is learning about `squeeze_scope` from a tooltip, and hasn't yet realised it is connected to `simp?`.

Reply:
Actually it won't work with `simp?`, at least as written currently. `squeeze_scope` will cause all enclosed `simp` calls to trace, but `simp?` is unaffected.